Changes to transition to generic API for ops with weight prepacking

kimishpatel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#35010 Changes to transition to generic API for ops with weight prepacking**

semantics.

Summary:
This PR moves all the xnnpack specific interfces to a generic interface.
Accordingly removes xnnpac specific reference from API and some variable
names.
What has not yet changed:

TODO:
USE_XNNPACK is still used. This can be removed where no XNNPACK
specific things are done. e.g., RegisterOpContext.cpp and
xnnpack_rewrite.cpp.
Also the filename and structure also remains. Some of the generic class
definition can be moved non-XNNPACK specific folder.

Test Plan:
python test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20526416](https://our.internmc.facebook.com/intern/diff/D20526416)

kimishpatel committed:
Changes to transition to generic API for ops with weight prepacking
semantics.

Summary:
This PR moves all the xnnpack specific interfces to a generic interface.
Accordingly removes xnnpac specific reference from API and some variable
names.
What has not yet changed:

TODO:
USE_XNNPACK is still used. This can be removed where no XNNPACK
specific things are done. e.g., RegisterOpContext.cpp and
xnnpack_rewrite.cpp.
Also the filename and structure also remains. Some of the generic class
definition can be moved non-XNNPACK specific folder.

Test Plan:
python test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kimishpatel commented:
To note: This PR has not yet done the following:
USE_XNNPACK is still used. This can be removed where no XNNPACK
specific things are done. e.g., RegisterOpContext.cpp and
xnnpack_rewrite.cpp.
Also the filename and structure also remains. Some of the generic class
definition can be moved non-XNNPACK specific folder.

dzhulgakov commented on aten/src/ATen/native/xnnpack/Convolution.cpp:
```diff
@@ -183,67 +137,111 @@ ContextConv2D create(
-        groups,
-        {},
-        {});
+Tensor run(const ContextConv2D& context, const Tensor& input) {
```
maybe minor thing - but why no make it a method?

kimishpatel replied:
I kept how it was in the original diff TBH. Yes it can be a method and in this case it might make sense. I will just keep it a TODO for future refactoring instead of doing it in the diff.

dzhulgakov commented on aten/src/ATen/native/xnnpack/Convolution.cpp:
```diff
@@ -183,67 +137,111 @@ ContextConv2D create(
+  return output.contiguous(input.suggest_memory_format());
+}
+
+c10::intrusive_ptr<xnnpack::Conv2dOpContext> Conv2dPrePack::operator()(
```
nit: you don't need this trampoline - you can register `xnnpack::XNNPackConv2dOpContext::create_context` directly with RegisterOps (it doesn't need to be a functor)

kimishpatel replied:
I tried that in the first diff itself and something I ran into that would not let me compile, cannot remember what. I will try again and see if there is a way around else report the error back here.

dzhulgakov commented on aten/src/ATen/native/xnnpack/OpContext.h:
```diff
@@ -2,95 +2,131 @@
+
+class LinearOpContext : public torch::jit::CustomClassHolder {
+ protected:
+  Tensor orig_weight_;
```
this duplicates all weights for serialization, I guess xnnpack doesn't have "unpack" method, does it?

kimishpatel replied:
This is correct. My plan was to have the weights set to null inside __setstate__ conditioned on some flags such as lite_interpreter run etc., that ensure we will not be serializing this. Is that the right approach?

kimishpatel replied:
@dzhulgakov, so I think we dont need to use trampoline. However, given that we are doing all this to make a generic API, it makes sense to hide backend specific context creation behind something, which in this PR turns out to be just a trampoline. However I am gonna stack a PR on top of this which will check global context to see which engine we want to use for packing (for now xnnpack) and direct context creation appropriately. I will refactor this code so that both __setstate__ method and PrePack op are directed to a common method for context creaton.

dreiss replied:
I'm not sure what the right approach is here.  If we set the weights to null, that means we can't re-save the model, right?

kimishpatel replied:
Yes, but if we know we will not need to, e.g. when we know we are doing inference in lite interpreter, it should ok, right?

dreiss replied:
Yeah, but what about people using the full JIT.  This is okay for now, but we should keep thinking about better solutions.

dzhulgakov replied:
Keeping the weights around is fine, we can gate freeing them with mobile interpreter flag indeed (we kind of do that for qnnpack on mobile I think).

dzhulgakov reviewed (commented):
Looks like on the right track!

Why do you have many formatting changes? Is it different version of clang-format?

kimishpatel commented:
> Why do you have many formatting changes? Is it different version of clang-format?

Possibly. I ended apply clang-format on all the files that I touched.

kimishpatel committed:
Update on "Changes to transition to generic API for ops with weight prepacking"


semantics.

Summary:
This PR moves all the xnnpack specific interfces to a generic interface.
Accordingly removes xnnpac specific reference from API and some variable
names.
What has not yet changed:

TODO:
USE_XNNPACK is still used. This can be removed where no XNNPACK
specific things are done. e.g., RegisterOpContext.cpp and
xnnpack_rewrite.cpp.
Also the filename and structure also remains. Some of the generic class
definition can be moved non-XNNPACK specific folder.

Test Plan:
python test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20526416](https://our.internmc.facebook.com/intern/diff/D20526416)

[ghstack-poisoned]

kimishpatel committed:
Update on "Changes to transition to generic API for ops with weight prepacking"


semantics.

Summary:
This PR moves all the xnnpack specific interfces to a generic interface.
Accordingly removes xnnpac specific reference from API and some variable
names.
What has not yet changed:

TODO:
USE_XNNPACK is still used. This can be removed where no XNNPACK
specific things are done. e.g., RegisterOpContext.cpp and
xnnpack_rewrite.cpp.
Also the filename and structure also remains. Some of the generic class
definition can be moved non-XNNPACK specific folder.

Test Plan:
python test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20526416](https://our.internmc.facebook.com/intern/diff/D20526416)

[ghstack-poisoned]

kimishpatel committed:
Update on "Changes to transition to generic API for ops with weight prepacking"


semantics.

Summary:
This PR moves all the xnnpack specific interfces to a generic interface.
Accordingly removes xnnpac specific reference from API and some variable
names.
What has not yet changed:

TODO:
USE_XNNPACK is still used. This can be removed where no XNNPACK
specific things are done. e.g., RegisterOpContext.cpp and
xnnpack_rewrite.cpp.
Also the filename and structure also remains. Some of the generic class
definition can be moved non-XNNPACK specific folder.

Test Plan:
python test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20526416](https://our.internmc.facebook.com/intern/diff/D20526416)

[ghstack-poisoned]

dreiss reviewed (changes requested):
Formatting changes need to be in a separate diff.  This is impossible to review asis.

kimishpatel commented:
> Formatting changes need to be in a separate diff. This is impossible to review asis.

I shouldnt have run clang-format on this. This is going to take a while to fix.

kimishpatel committed:
Update on "Changes to transition to generic API for ops with weight prepacking"


semantics.

Summary:
This PR moves all the xnnpack specific interfces to a generic interface.
Accordingly removes xnnpac specific reference from API and some variable
names.
What has not yet changed:

TODO:
USE_XNNPACK is still used. This can be removed where no XNNPACK
specific things are done. e.g., RegisterOpContext.cpp and
xnnpack_rewrite.cpp.
Also the filename and structure also remains. Some of the generic class
definition can be moved non-XNNPACK specific folder.

Test Plan:
python test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20526416](https://our.internmc.facebook.com/intern/diff/D20526416)

[ghstack-poisoned]

kimishpatel committed:
Update on "Changes to transition to generic API for ops with weight prepacking"


semantics.

Summary:
This PR moves all the xnnpack specific interfces to a generic interface.
Accordingly removes xnnpac specific reference from API and some variable
names.
What has not yet changed:

TODO:
USE_XNNPACK is still used. This can be removed where no XNNPACK
specific things are done. e.g., RegisterOpContext.cpp and
xnnpack_rewrite.cpp.
Also the filename and structure also remains. Some of the generic class
definition can be moved non-XNNPACK specific folder.

Test Plan:
python test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20526416](https://our.internmc.facebook.com/intern/diff/D20526416)

[ghstack-poisoned]

kimishpatel committed:
Update on "Changes to transition to generic API for ops with weight prepacking"


semantics.

Summary:
This PR moves all the xnnpack specific interfces to a generic interface.
Accordingly removes xnnpac specific reference from API and some variable
names.
What has not yet changed:

TODO:
USE_XNNPACK is still used. This can be removed where no XNNPACK
specific things are done. e.g., RegisterOpContext.cpp and
xnnpack_rewrite.cpp.
Also the filename and structure also remains. Some of the generic class
definition can be moved non-XNNPACK specific folder.

Test Plan:
python test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20526416](https://our.internmc.facebook.com/intern/diff/D20526416)

[ghstack-poisoned]

kimishpatel committed:
Update on "Changes to transition to generic API for ops with weight prepacking"


semantics.

Summary:
This PR moves all the xnnpack specific interfces to a generic interface.
Accordingly removes xnnpac specific reference from API and some variable
names.
What has not yet changed:

TODO:
USE_XNNPACK is still used. This can be removed where no XNNPACK
specific things are done. e.g., RegisterOpContext.cpp and
xnnpack_rewrite.cpp.
Also the filename and structure also remains. Some of the generic class
definition can be moved non-XNNPACK specific folder.

Test Plan:
python test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20526416](https://our.internmc.facebook.com/intern/diff/D20526416)

[ghstack-poisoned]

kimishpatel committed:
Update on "Changes to transition to generic API for ops with weight prepacking"


semantics.

Summary:
This PR moves all the xnnpack specific interfces to a generic interface.
Accordingly removes xnnpac specific reference from API and some variable
names.
What has not yet changed:

TODO:
USE_XNNPACK is still used. This can be removed where no XNNPACK
specific things are done. e.g., RegisterOpContext.cpp and
xnnpack_rewrite.cpp.
Also the filename and structure also remains. Some of the generic class
definition can be moved non-XNNPACK specific folder.

Test Plan:
python test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20526416](https://our.internmc.facebook.com/intern/diff/D20526416)

[ghstack-poisoned]

kimishpatel committed:
Update on "Changes to transition to generic API for ops with weight prepacking"


semantics.

Summary:
This PR moves all the xnnpack specific interfces to a generic interface.
Accordingly removes xnnpac specific reference from API and some variable
names.
What has not yet changed:

TODO:
USE_XNNPACK is still used. This can be removed where no XNNPACK
specific things are done. e.g., RegisterOpContext.cpp and
xnnpack_rewrite.cpp.
Also the filename and structure also remains. Some of the generic class
definition can be moved non-XNNPACK specific folder.

Test Plan:
python test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20526416](https://our.internmc.facebook.com/intern/diff/D20526416)

[ghstack-poisoned]

kimishpatel committed:
Update on "Changes to transition to generic API for ops with weight prepacking"


semantics.

Summary:
This PR moves all the xnnpack specific interfces to a generic interface.
Accordingly removes xnnpac specific reference from API and some variable
names.
What has not yet changed:

TODO:
USE_XNNPACK is still used. This can be removed where no XNNPACK
specific things are done. e.g., RegisterOpContext.cpp and
xnnpack_rewrite.cpp.
Also the filename and structure also remains. Some of the generic class
definition can be moved non-XNNPACK specific folder.

Test Plan:
python test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20526416](https://our.internmc.facebook.com/intern/diff/D20526416)

[ghstack-poisoned]

kimishpatel committed:
Update on "Changes to transition to generic API for ops with weight prepacking"


semantics.

Summary:
This PR moves all the xnnpack specific interfces to a generic interface.
Accordingly removes xnnpac specific reference from API and some variable
names.
What has not yet changed:

TODO:
USE_XNNPACK is still used. This can be removed where no XNNPACK
specific things are done. e.g., RegisterOpContext.cpp and
xnnpack_rewrite.cpp.
Also the filename and structure also remains. Some of the generic class
definition can be moved non-XNNPACK specific folder.

Test Plan:
python test/test_xnnpack_integration.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20526416](https://our.internmc.facebook.com/intern/diff/D20526416)

[ghstack-poisoned]

jerryzh168 commented:
What's the plan? are we going to use this for quantized ops as well?

kimishpatel commented:
> What's the plan? are we going to use this for quantized ops as well?

I don't think so. Changes are mainly to abstract away xnnpack specific namespace from ops and make it generic. For example if we want to use gpu as the backend instead of xnnpack.

dreiss commented:
> What's the plan? are we going to use this for quantized ops as well?

Not my plan.  Quantized ops need to have output scale/zero as part of the interface, so IMO they should follow this same pattern but be separate.

jerryzh168 commented:
folding quantized prepack is pretty much the same I think: https://github.com/pytorch/pytorch/pull/35077

So is this for the non quantized use case only?

jerryzh168 commented:
Oh sorry, it might be confusing, I'm just talking about prepack folding. Found this PR when after I rebase my PRs and see the name of `FoldPrepackingOps` is changed

kimishpatel commented:
> Oh sorry, it might be confusing, I'm just talking about prepack folding. Found this PR when after I rebase my PRs and see the name of `FoldPrepackingOps` is changed

Not sure what you mean by FoldPrepackingOps name changed. Also the rest of the op registration namespace changed but that should not impact any JIT passes.

jerryzh168 commented on torch/csrc/jit/passes/xnnpack_rewrite.cpp:
```diff
@@ -68,58 +72,62 @@ void insertXNNPACKConv2dOp(std::shared_ptr<Graph>& graph) {
+        n->kind() == Symbol::fromQualString("prepacked::conv2d_clamp_prepack"));
   };
-  FoldPrePackingOps(m, filter_fn, "xnnpack_prepack_folding");
+  PrePackingOpsFolder(m, filter_fn, "prepack_folding");
```
I'm talking about the function name is changed from `FoldPrePackingOps` to `PrePackingOpsFolder`, which leads me to the PR, but it is not important. I'm just trying to understand what is the use case this function is targeting, is this only for non-quantized prepacking ops?

kimishpatel replied:
Aah thanks for pointing out :). Shouldn't have changed that as part of this PR. But to answer your question, no. It should be generic as we had discussed originally when the PR for this pass was under review. It should still be a generally usable pass that is can be applied after freezing, regardless. However the naming either `FoldPrePackingOps` or `PrePackingOpsFolder` does suggest that it is for prepacking ops, including quantized. But I understand the concern, if that is the concern, that the naming may suggest that it is applied to ops that are in `prepacked` namespace even though that is not the case.

jerryzh168 replied:
Yeah the implementation is applicable to quantized ops as well. Do we want to do use this for quantized prepacking ops as well?

kimishpatel replied:
I think that was the original plan, right? I dont think anything about that has changed, except naming. So I would say yes we should.

jerryzh168 replied:
I mean yeah, I can use `PrePackingOpsFolder` to implement a pass to fold quantized ops.
My question is should we merge that to the current new `FoldPrePackingOps` here, or is this still function still just for xnnpack?

jerryzh168 replied:
resolved in chat, this is prepack folding for fp32 ops.
